go/vt/mysqlctl/backup.go

```diff
@@ -0,0 +1,563 @@
+	}
+	return fd, nil
+}
+
+// BackupManifest represents the backup. It lists all the files, and
+// the ReplicationPosition that the backup was taken at.
+type BackupManifest struct {
+	// FileEntries contains all the files in the backup
+	FileEntries []FileEntry
+
+	// ReplicationPosition is the position at which the backup was taken
+	ReplicationPosition proto.ReplicationPosition
+}
+
+// isDbDir returns true if the given directory contains a DB
+func isDbDir(p string) bool {
+	// db.opt is there
+	if _, err := os.Stat(path.Join(p, "db.opt")); err == nil {
+		return true
+	}
+
+	// Look for at least one .frm file
+	fis, err := ioutil.ReadDir(p)
+	if err != nil {
+		return false
+	}
+	for _, fi := range fis {
+		if strings.HasSuffix(fi.Name(), ".frm") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func addDirectory(fes []FileEntry, base string, baseDir string, subDir string) ([]FileEntry, error) {
+	p := path.Join(baseDir, subDir)
+
+	fis, err := ioutil.ReadDir(p)
+	if err != nil {
+		return nil, err
+	}
+	for _, fi := range fis {
+		fes = append(fes, FileEntry{
+			Base: base,
+			Name: path.Join(subDir, fi.Name()),
+		})
+	}
+	return fes, nil
+}
+
+func findFilesTobackup(cnf *Mycnf, logger logutil.Logger) ([]FileEntry, error) {
+	var err error
+	var result []FileEntry
+
+	// first add inno db files
+	result, err = addDirectory(result, backupInnodbDataHomeDir, cnf.InnodbDataHomeDir, "")
+	if err != nil {
+		return nil, err
+	}
+	result, err = addDirectory(result, backupInnodbLogGroupHomeDir, cnf.InnodbLogGroupHomeDir, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// then add DB directories
+	fis, err := ioutil.ReadDir(cnf.DataDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, fi := range fis {
+		p := path.Join(cnf.DataDir, fi.Name())
+
+		// If this is not a directory, try to eval it as a syslink.
+		if !fi.IsDir() {
+			p, err = filepath.EvalSymlinks(p)
+			if err != nil {
+				return nil, err
+			}
+			fi, err = os.Stat(p)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if isDbDir(p) {
+			result, err = addDirectory(result, backupData, cnf.DataDir, fi.Name())
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return result, nil
+}
+
+// Backup is the main entry point for a backup:
+// - uses the BackupStorage service to store a new backup
+// - shuts down Mysqld during the backup
+// - remember if we were replicating, restore the exact same state
+func Backup(mysqld MysqlDaemon, logger logutil.Logger, bucket, name string, backupConcurrency int, hookExtraEnv map[string]string) error {
+
+	// start the backup with the BackupStorage
+	bs := backupstorage.GetBackupStorage()
+	bh, err := bs.StartBackup(bucket, name)
+	if err != nil {
+		return fmt.Errorf("StartBackup failed: %v", err)
+	}
+
+	if err = backup(mysqld, logger, bh, backupConcurrency, hookExtraEnv); err != nil {
+		if err := bh.AbortBackup(); err != nil {
+			logger.Errorf("failed to abort backup: %v", err)
+		}
```

Comment:
By the letter of the style guide, you should have a "return err" here, so the happy path below can be unindented ("return bh.EndBackup()"). The err from AbortBackup shouldn't interfere since it's scoped to the "if" block, but you could rename that abortErr if you're worried it'll be confusing.

It's not really a clear readability win though in this case, so up to you.